File: guests/haskell/stall.hs
-- The stall scene, Haskell port — guests/rust/stall.rs,
-- tools/scenes/stall.steps.

import Control.Concurrent (threadDelay)
import KayaApp
import KayaWire (Value (..))

-- Past the watchdog's one-second threshold; threadDelay counts microseconds.
blockMicros :: Int
blockMicros = 2500000

-- A day, never a literal park (docs/traps.md, the stall scene wedges for a DAY).
wedgeMicros :: Int
wedgeMicros = 86400 * 1000000

main :: IO ()
main = kayaMain $ \app -> do
  _ <- buildTx app $ do
    window 0 [WTitle "stall"]
    status <- signal (VStr "ready")

    root <-
      column
        []
        [ labelBound status [A11yId "status"], -- label#0
          -- DELIBERATELY WRONG, and the only place in this repo that is.
          buttonOn
            "block" -- button#0
            (threadDelay blockMicros)
            [],
          buttonOn
            "ping" -- button#1
            (buildTx app (writeSignal status (VStr "pinged")))
            [],
          buttonOn
            "wedge" -- button#2
            (threadDelay wedgeMicros)
            []
        ]
    mount root
    return ()
  return ()

File: guests/rust/stall.rs
//! THE ONE GUEST THAT MISUSES KAYA ON PURPOSE (tools/scenes/stall.steps):
//! do not "fix" the sleeps onto a worker; `ping` makes the work PENDING.

pub(crate) fn app(ctx: kaya::AppCtx) {
    #[derive(Clone, Copy)]
    enum Msg {
        Block,
        Ping,
        Wedge,
    }

    // Past the watchdog's one-second threshold, and no longer.
    const BLOCK: std::time::Duration = std::time::Duration::from_millis(2500);

    // A day, never a literal park (docs/traps.md, "The stall scene wedges
    // for a DAY").
    const WEDGE: std::time::Duration = std::time::Duration::from_secs(86_400);

    let msgs = kaya::Messages::<Msg>::new();
    let status = ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW).title("stall");
        let status = tx.signal("ready");
        let root = tx
            .column(|tx| {
                tx.label(status); // label#0
                let block = tx.button("block").id(); // button#0
                msgs.on_click(block, Msg::Block);
                let ping = tx.button("ping").id(); // button#1
                msgs.on_click(ping, Msg::Ping);
                let wedge = tx.button("wedge").id(); // button#2
                msgs.on_click(wedge, Msg::Wedge);
            })
            .id();
        tx.mount(root);
        status
    });

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            // DELIBERATELY WRONG, and the only place in this repo that is.
            Msg::Block => std::thread::sleep(BLOCK),
            Msg::Ping => ctx.apply(|tx| tx.write(status, "pinged")),
            Msg::Wedge => std::thread::sleep(WEDGE),
        }
    }
}

fn main() {
    kaya::run(app)
}
